File: 프로그래머스/3/138475. 억억단을 외우자/억억단을 외우자.py
def solution(e, starts):
    spf = [0] * (e + 1)
    exp = [0] * (e + 1)
    cnt = [0] * (e + 1)
    cnt[1] = 1

    for i in range(2, e + 1):
        if spf[i] == 0:
            spf[i] = i
            exp[i] = 2
            cnt[i] = 2
            for j in range(i * 2, e + 1, i):
                if spf[j] == 0:
                    spf[j] = i
        else:
            prev = i // spf[i]
            if spf[prev] == spf[i]:
                exp[i] = exp[prev] + 1
                cnt[i] = cnt[prev] // exp[prev] * exp[i]
            else:
                exp[i] = 2
                cnt[i] = cnt[prev] * 2

    max_cnt = 0
    num = [0] * (e + 1)
    for i in range(e, 0, -1):
        if cnt[i] >= max_cnt:
            max_cnt = cnt[i]
            num[i] = i
        else:
            num[i] = num[i + 1]

    return [num[s] for s in starts]

File: 프로그래머스/3/138475. 억억단을 외우자/억억단을 외우자.java
class Solution {
    public int[] solution(int e, int[] starts) {
        int i;
        int j;
        int len;
        int max;
        int prev;
        int[] dp;
        int[] pf;
        int[] num;
        int[] ans;
        int[] pfExp;

        dp = new int[e + 1];
        pf = new int[e + 1];
        pfExp = new int[e + 1];
        dp[1] = 1;
        for (i = 2; i <= e; i++) {
            if (pf[i] == 0) {
                pf[i] = i;
                pfExp[i] = 2;
                dp[i] = 2;
                for (j = i << 1; j <= e; j += i) {
                    if (pf[j] == 0) {
                        pf[j] = i;
                    }
                }
            } else {
                prev = i / pf[i];
                if (pf[prev] == pf[i]) {
                    pfExp[i] = pfExp[prev] + 1;
                    dp[i] = dp[prev] / pfExp[prev] * pfExp[i];
                } else {
                    pfExp[i] = 2;
                    dp[i] = dp[prev] << 1;
                }
            }
        }
        max = 0;
        num = new int[e + 1];
        for (i = e; i >= 1; i--) {
            if (dp[i] >= max) {
                max = dp[i];
                num[i] = i;
            } else {
                num[i] = num[i + 1];
            }
        }
        len = starts.length;
        ans = new int[len];
        for (i = 0; i < len; i++) {
            ans[i] = num[starts[i]];
        }
        return ans;
    }
}